File: SETool_Python_public/process_submission.py
"""
Process user submissions and generate results
"""

import os
import shutil
import zipfile
import numpy as np
import pandas as pd
from datetime import datetime

from email_utils import (send_submission_received, send_results, 
                        send_duplicate_notification, send_incorrect_submission)
from validate_submission import validate_submission
from obtain_output_data import obtain_output_data
from create_figures import create_figures

def process_submission(filename, filenum, data, output_folder, file_folder, 
                       rootfolder, config):
    """
    Process a model submission
    
    Parameters:
    -----------
    filename : str
        Path to the submitted .zip file
    filenum : int
        Submission ID number
    data : dict
        Test data for all vehicle masses
    output_folder : str
        Temporary folder for processing
    file_folder : str
        Main folder containing submissions
    rootfolder : str
        Root directory of the tool
    config : Config
        Configuration object with all settings
    """
    
    # Extract zip file
    try:
        with zipfile.ZipFile(filename, 'r') as zip_ref:
            zip_ref.extractall(output_folder)
        print('Folder detected and .zip extracted, now deleting .zip file.')
        os.remove(filename)
    except Exception as e:
        print(f'Error extracting zip file: {str(e)}')
        return
    
    # Create a copy in Models Saved
    saved_folder = os.path.join(rootfolder, 'Models Saved')
    data_filename = os.path.join(saved_folder, f'Model__{filenum}.zip')
    shutil.make_archive(data_filename.replace('.zip', ''), 'zip', output_folder)
    
    # Add output folder to path
    import sys
    sys.path.insert(0, output_folder)
    
    # Extract settings and model file information
    xlsx_files = [f for f in os.listdir(output_folder) if f.endswith('.xlsx')]
    
    if not xlsx_files:
        print("No .xlsx file found in folder, no e-mail could be sent!")
        os.remove(data_filename)
        return
    
    # Read settings
    settings_file = os.path.join(output_folder, xlsx_files[0])
    try:
        settings_data = pd.read_excel(settings_file, header=None)
        
        if settings_data.shape[0] < 4 or settings_data.shape[1] < 2:
            print(".xlsx file has the wrong format, no e-mail could be sent!")
            os.remove(data_filename)
            return
        
        author_name = settings_data.iloc[0, 1]
        author_affiliation = settings_data.iloc[1, 1]
        author_email = settings_data.iloc[2, 1]
        model_name = settings_data.iloc[3, 1]
        
    except Exception as e:
        print(f"Error reading settings file: {str(e)}")
        os.remove(data_filename)
        return
    
    # Check for Model file (Python or MATLAB)
    model_file_py = os.path.join(output_folder, 'Model.py')
    model_file_m = os.path.join(output_folder, 'Model.m')
    model_file_p = os.path.join(output_folder, 'Model.p')
    
    has_model = (os.path.exists(model_file_py) or 
                 os.path.exists(model_file_m) or 
                 os.path.exists(model_file_p))
    
    if not has_model:
        print("No Model file found in folder, email sent!")
        send_incorrect_submission(author_email, model_name)
        os.remove(data_filename)
        return
    
    # Determine model type
    if os.path.exists(model_file_py):
        model_type = "Python"
    else:
        model_type = "MATLAB"
    
    print(f"Model file found! Type: {model_type}")
    
    # Send confirmation email
    send_submission_received(author_email, model_name)
    print("Model file found! Email sent that model was received correctly.")
    
    # Validate whether submission can be run without error and doesn't exceed computation time
    fail = validate_submission(data, config.inputs, author_email, model_type)
    if fail:
        os.remove(data_filename)
        print("TESTING TESTING 1")
        return
    
    # Create all output data using Model
    output_data, rmse, maxe, mae, rmse_charge, file_data, complexity = \
        obtain_output_data(config.setups, data, config.inputs)
    
    # Check for possible exploitation
    if np.mean(rmse) > 15:
        exploit = True
        print("\033[91mWARNING: Possible exploitation detected, no data will be sent to user, please check output data!!!\033[0m")
        # Rename saved file to indicate exploit
        exploit_filename = data_filename.replace('.zip', '_EXPLOIT.zip')
        os.rename(data_filename, exploit_filename)
    else:
        exploit = False
    
    # Create all test figures and compute robustness errors
    mean_rmse_temp, rmse_soc_offset, rmse_sens_offset = \
        create_figures(output_data, config.inputs, data, file_data, 
                      output_folder, rmse, config.temperatures)
    
    # Calculate average RMSE for all specified test case results
    rmse_results = calculate_rmse_results(rmse, rmse_charge, config)
    
    # Weight SOC offset errors
    rmse_soc_weighted = weight_soc_offset_errors(rmse_soc_offset)
    
    # Get relevant sensor offset errors (ignoring 0.05A offsets since these are very small)
    rmse_sens_offset_relevant = [rmse_sens_offset[i] for i in [0, 5, 6, 11, 12, 17]]
    
    # Create bar plot figure for average RMSE per test case
    create_summary_barplot(rmse_results, config.tests, output_folder)
    
    # Determine ranking score and weightings
    scores = calculate_scores(rmse_results, mean_rmse_temp, rmse_soc_weighted, 
                             rmse_sens_offset_relevant)
    
    # Load table of ranking scores and weightings
    weighted_score = create_weighted_score_table(config.tests_weighted, 
                                                  config.weights, scores)
    final_score = np.sum(weighted_score['Weighted_Value'])
    
    # Save figures and error summary
    save_results(output_data, output_folder, saved_folder, filenum, config.email_size_limit)
    
    os.chdir(rootfolder)
    
    # Create leaderboard entry
    stats = create_leaderboard_entry(datetime.now(), author_name, author_affiliation,
                                     model_name, final_score, scores, rmse_results,
                                     mae, maxe, complexity, filenum)
    
    # Update leaderboard
    leaderboard_path = os.path.join(file_folder, 'Leaderboard.csv')
    rank, total, is_duplicate = update_leaderboard(leaderboard_path, stats)
    
    if is_duplicate:
        send_duplicate_notification(author_email, model_name)
        duplicate_filename = data_filename.replace('.zip', '_DUPLICATE.zip')
        os.rename(data_filename, duplicate_filename)
        print('\033[91mSubmission has identical test results to an existing entry. '
              'Ignoring duplicate submission!\033[0m')
        return
    
    # Send results to author
    data_files = get_result_files(saved_folder, filenum, config.email_size_limit)
    send_results(author_email, model_name, rank, total, leaderboard_path, 
                data_files, exploit)
    
    if not fail:
        print(f'The Model has been processed and ranks as #{rank} on the leaderboard.')
    
    # Cleanup
    sys.path.remove(output_folder)

# Helper functions

def calculate_rmse_results(rmse, rmse_charge, config):
    """Calculate RMSE results for all test cases"""
    rmse_results = np.zeros(10)
    rmse_means = [np.mean(rmse[:, i]) for i in range(4)]
    
    rmse_results[0] = np.mean(rmse[rmse != 0])
    rmse_results[1] = rmse_means[1]
    rmse_results[2] = np.mean([rmse_means[0], rmse_means[2], rmse_means[3]])
    rmse_results[3] = np.mean(rmse_charge[rmse_charge != 0])
    rmse_results[4:8] = rmse_means
    
    # Separate standard and custom cycles
    standard_rmse = []
    custom_rmse = []
    total_cycles = config.normal_cycles + config.custom_cycles
    
    for i in range(0, len(rmse), total_cycles):
        standard_rmse.append(rmse[i:i+config.normal_cycles])
        custom_rmse.append(rmse[i+config.normal_cycles:i+total_cycles])
    
    standard_rmse = np.concatenate(standard_rmse)
    custom_rmse = np.concatenate(custom_rmse)
    
    rmse_results[8] = np.mean(standard_rmse[standard_rmse != 0])
    rmse_results[9] = np.mean(custom_rmse[custom_rmse != 0])
    
    return rmse_results

def weight_soc_offset_errors(rmse_soc_offset):
    """Weigh the values according to the datasize that produced it (90% iSOC >> 30% iSOC)"""
    rmse_soc_weighted = []
    
    for i in range(0, len(rmse_soc_offset), 3):
        for j in range(3):
            replication_factor = 3 - j  # Calculate the replication factor (3, 2, 1 for each j)
            rmse_soc_weighted.extend([rmse_soc_offset[i+j]] * replication_factor)
    
    return rmse_soc_weighted

def calculate_scores(rmse_results, mean_rmse_temp, rmse_soc_weighted, 
                     rmse_sens_offset_relevant):
    """Calculate all test scores"""
    scores = list(rmse_results)
    scores.extend(mean_rmse_temp)
    scores.append(np.mean(rmse_soc_weighted))
    scores.append(np.mean(rmse_sens_offset_relevant))
    return scores

def create_weighted_score_table(test_names, weights, scores):
    """Create a pandas DataFrame with weighted scores"""
    weighted_values = [w * s for w, s in zip(weights, scores)]
    
    df = pd.DataFrame({
        'Test_Case': test_names,
        'Weights': weights,
        'Scores': scores,
        'Weighted_Value': weighted_values
    })
    
    return df

def create_summary_barplot(rmse_results, tests, output_folder):
    """Create bar plot summary of RMSE results"""
    import matplotlib.pyplot as plt
    
    fig, ax = plt.subplots(figsize=(12, 6))
    bars = ax.bar(range(len(rmse_results)), rmse_results)
    ax.set_xticks(range(len(rmse_results)))
    ax.set_xticklabels(tests, rotation=45, ha='right')
    ax.set_ylabel('SOC Estimation Average RMS Error (%)')
    ax.grid(True, alpha=0.3)
    ax.set_title('Test 1 to 8 RMS Error')
    
    # Add value labels on bars
    for i, (bar, val) in enumerate(zip(bars, rmse_results)):
        height = bar.get_height()
        ax.text(bar.get_x() + bar.get_width()/2., height,
                f'{val:.2f}%', ha='center', va='bottom', fontsize=8)
    
    plt.tight_layout()
    plt.savefig(os.path.join(output_folder, 'Test_01_to_08_RMSE.png'), dpi=150)
    plt.close()

def save_results(output_data, output_folder, saved_folder, filenum, size_limit):
    """Save error summary and figures"""
    import pickle
    
    # Save output data
    summary_file = os.path.join(output_folder, 'Error_Summary_Table.pkl')
    with open(summary_file, 'wb') as f:
        pickle.dump(output_data, f)
    
    # Create zip file
    zip_filename = os.path.join(saved_folder, f'Model_{filenum} Error Summary.zip')
    
    with zipfile.ZipFile(zip_filename, 'w') as zipf:
        zipf.write(summary_file, 'Error_Summary_Table.pkl')
        for file in os.listdir(output_folder):
            if file.endswith('.png'):
                zipf.write(os.path.join(output_folder, file), file)
    
    # Check size and split if necessary
    file_size_mb = os.path.getsize(zip_filename) / (1024 * 1024)
    
    if file_size_mb > size_limit:
        # Split into two files
        zip_filename1 = os.path.join(saved_folder, 
                                     f'Model_{filenum} Error Summary 1 of 2.zip')
        zip_filename2 = os.path.join(saved_folder, 
                                     f'Model_{filenum} Error Summary 2 of 2.zip')
        
        with zipfile.ZipFile(zip_filename1, 'w') as zipf:
            zipf.write(summary_file, 'Error_Summary_Table.pkl')
        
        with zipfile.ZipFile(zip_filename2, 'w') as zipf:
            for file in os.listdir(output_folder):
                if file.endswith('.png'):
                    zipf.write(os.path.join(output_folder, file), file)
        
        os.remove(zip_filename)
        print('Figures were too large to send a single mail, the data was split up instead!')
    
    # Cleanup
    for file in os.listdir(output_folder):
        if file.endswith('.png'):
            os.remove(os.path.join(output_folder, file))
    os.remove(summary_file)

def create_leaderboard_entry(dt, author_name, affiliation, model_name, 
                             final_score, scores, rmse_results, mae, maxe, 
                             complexity, filenum):
    """Create a new leaderboard entry"""
    stats = {
        'Submission_Time': dt.strftime('%Y-%m-%d %H:%M:%S'),
        'Author': author_name,
        'Affiliation': affiliation,
        'Model_Name': model_name,
        'Weighted_Error': round(final_score, 3)
    }
    
    # Add all test scores
    for i, score in enumerate(scores):
        stats[f'Score_{i+1}'] = round(score, 3)
    
    stats['All_Drive_Cycles_Average_RMSE'] = round(rmse_results[0], 3)
    stats['All_Drive_Cycles_Average_MAE'] = round(np.mean(mae[mae != 0]), 3)
    stats['All_Drive_Cycles_Average_MAXE'] = round(np.mean(maxe[maxe != 0]), 3)
    stats['Complexity'] = complexity
    stats['Submission_ID'] = filenum
    
    return stats

def update_leaderboard(leaderboard_path, stats):
    """Update the leaderboard with new entry"""
    if os.path.exists(leaderboard_path):
        df = pd.read_csv(leaderboard_path)
        
        # Check for duplicates
        score_cols = [col for col in df.columns if col.startswith('Score_')]
        new_scores = np.array([stats[col] for col in score_cols])
        existing_scores = df[score_cols].values
        
        is_duplicate = np.any(np.all(existing_scores == new_scores, axis=1))
        
        if is_duplicate:
            return None, len(df), True
        
        # Add new entry
        new_row = pd.DataFrame([stats])
        df = pd.concat([df, new_row], ignore_index=True)
        
        # Sort by weighted error
        df = df.sort_values(['Weighted_Error', 'Submission_Time'])
        df['Position'] = range(1, len(df) + 1)
        
        # Find rank of new entry
        rank = df[df['Submission_ID'] == stats['Submission_ID']]['Position'].values[0]
        
    else:
        df = pd.DataFrame([stats])
        df['Position'] = 1
        rank = 1
    
    df.to_csv(leaderboard_path, index=False)
    
    return rank, len(df), False

def get_result_files(saved_folder, filenum, size_limit):
    """Get list of result files to attach to email"""
    single_file = os.path.join(saved_folder, f'Model_{filenum} Error Summary.zip')
    
    if os.path.exists(single_file):
        return single_file
    else:
        file1 = os.path.join(saved_folder, f'Model_{filenum} Error Summary 1 of 2.zip')
        file2 = os.path.join(saved_folder, f'Model_{filenum} Error Summary 2 of 2.zip')
        return [file1, file2]

File: Legacy App/src/webserver/webserver/mlSource/Process_Submission.m
function Process_Submission(filename, filenum, Data, outputFolder, fileFolder, rootfolder, Inputs, Setups, Normal_cycles, Custom_cycles , Tests, Tests_weighted,Temperatures)  
    unzip(filename,outputFolder);                                   % Unzip .zip into temporary folder
    fprintf('Folder detected and .zip extracted, now deleting .zip file.\n');        
    delete(filename);                                               % Delete extracted .zip model folder 
    clear filename;
    DataFileName = fullfile(strcat(fileFolder,' Saved'), sprintf('Model__%d.zip',filenum)); % Create folder copy name
    zip(DataFileName,outputFolder);                                 % Zip copy into /Models Saved
    addpath(strcat(fileFolder,'\Blind Model'))                      % Add extracted folder to path
    %% Extracting settings and .m/.p file information
    filepath = outputFolder;                   % Access temporary folder location
    if ~isempty(dir(strcat(filepath,'\*.xlsx')))                    % If there exists a *.xlsx file with the correct format
        filename = dir(strcat(filepath,'\*.xlsx')).name;            % Find information associated to the .xlsx
        Settings_Data = readcell(strcat(filepath,'\',filename));    % Save settings data for mail
        if size(Settings_Data,1) == 4 && size(Settings_Data,2) == 2 % Check size of settings
        Author_Name = Settings_Data(1,2);
        Author_Affiliation = Settings_Data(2,2);
        Author_Email = Settings_Data(3,2);
        Model_Name = Settings_Data(4,2);
        if(isfile(strcat(filepath,'\Model.m')))||isfile(strcat(filepath,'\Model.p')) % Only continue if there is a Model.m or .p file
        title1_temp = {'Your model named "', Model_Name{1},'" has been received!'}; % Add email title
        Title1 = join(title1_temp);
        sendmail(Author_Email{1},Title1{1}, ...                 % Write and send email for submission confirmation
          ['Dear Author, Your data has been received and is being processed. An error summary of your results will be sent to you shortly, typically within ten minutes or less. If an error occurs, you will be notified. ' ...
          'If you have not received an email within 3h, please check your spam folder or contact us. Thank you for your submission!']);
        fprintf("Model.m or Model.p file found! Email sent that model was received correctly.\n")
        %% Validate whether submission can be run without error and doesn't exceed computation time, see "Validate_Submission.m"
        fail = Validate_Submission(Data, Author_Email);
        if fail == 1
            delete(DataFileName)
            return
        end
        %% Create all output data using Model
        [OutputData, RMSE, MAXE, MAE, RMSE_Charge, File, Complexity] = Obtain_Output_Data(Setups,Data,Inputs);
        
        %% Check for possible exploitation
        if mean(mean(RMSE)) > 15
            exploit = 1;
            fprintf(2,"WARNING: Possible exploitation detected, no data will be sent to user, please check output data!!!\n")
            temp = dir().folder;
            tempFolder = strcat(temp,"\Models Saved");
            movefile(fullfile(tempFolder,strcat('Model__',mat2str(filenum),'.zip')), ...
            fullfile(tempFolder,strcat('Model__',mat2str(filenum),'_EXPLOIT.zip')));
            clear temp; clear tempFolder;
        else
            exploit = 0;
        end
        %% Create all test figures and compute robustness errors
        [mean_RMSE_temp, RMSE_SOC_Offset, RMSE_Sens_Offset] = Create_Figures(OutputData,Inputs, Data, File, outputFolder,RMSE,Temperatures);

        %% Obtain average RMSE for all specified test case results
        RMSE_Results = zeros(10,1);
        RMSE_means = [mean(RMSE(:,1)),mean(RMSE(:,2)),mean(RMSE(:,3)),mean(RMSE(:,4))];
        RMSE_Results(1) = mean(nonzeros(RMSE));
        RMSE_Results(2) = RMSE_means(2);
        RMSE_Results(3) = mean([RMSE_means(1),RMSE_means(3),RMSE_means(4)]);
        RMSE_Results(4) = mean(nonzeros(RMSE_Charge));
        RMSE_Results(5:8) = RMSE_means(1:4);
        Standard_RMSE = [];
        Custom_RMSE = [];
        for i = 1:(Normal_cycles+Custom_cycles):length(RMSE)
            Standard_RMSE = [Standard_RMSE; RMSE(i:i+3,:)];
            Custom_RMSE = [Custom_RMSE; RMSE(i+4:i+5,:)];
        end
        Standard_RMSE_means = [mean(Standard_RMSE(:,1)),mean(Standard_RMSE(Standard_RMSE(:,2)~=0,2)),mean(Standard_RMSE(:,3)),mean(Standard_RMSE(:,4))];
        Custom_RMSE_means = [mean(Custom_RMSE(:,1)),mean(Custom_RMSE(Custom_RMSE(:,2)~=0,2)),mean(Custom_RMSE(:,3)),mean(Custom_RMSE(:,4))];
        RMSE_Results(9) = mean(Standard_RMSE_means);
        RMSE_Results(10) = mean(Custom_RMSE_means);
        RMSE_SOC_weighted = [];                                     % Weigh the values according to the datasize that produced it (90% iSOC >> 30% iSOC)
        for i = 0:3:length(RMSE_SOC_Offset)-3
            for j = 1:3
                replication_factor = 3 - (j - 1);                   % Calculate the replication factor (3, 2, 1 for each j)
                RMSE_SOC_weighted = [RMSE_SOC_weighted; repmat(RMSE_SOC_Offset(i+j), replication_factor, 1)];
            end
        end
        RMSE_Sens_Offset_relevant = RMSE_Sens_Offset([1 6 7 12 13 18]); % Ignore the 0.05A offsets since these are very small
        clear Standard_RMSE; clear Custom_RMSE;
        %% Create barplot figure for average RMSE per test case
        figure;
        b = bar(RMSE_Results);                                      % Create barplot
        set(gca, 'XTickLabel', Tests);                              % Name bars
        xtips1 = b(1).XEndPoints;
        ytips1 = b(1).YEndPoints;
        labels1 = strcat(compose("%.2f",b(1).YData),"\%");  
        text(xtips1,ytips1,labels1,'HorizontalAlignment','center','VerticalAlignment','bottom') % Add values on top of bars
        ylabel('SOC Estimation Average RMS Error (\%)', 'Interpreter','latex')
        grid on
        title('Test 1 to 8 RMS Error')

        savefig('Test 01 to 08 - RMSE (Bar plot)')                  % Save figure
        close
        %% Determine ranking score and weightings
        weightedScore = table('Size', [length(Tests_weighted),4], 'VariableTypes', {'string', 'single', 'single', 'single'},'VariableNames', {'Test_Case', 'Weights','Scores', 'Weighted_Value'});
        Weights = {0 1/10 1/10 1/10 1/30 2/30 2/30 1/30 1/10 1/10 1/60 1/60 1/60 1/60 1/60 1/60 1/10 1/10}'; % equal distribution per testcase
        Scores = {RMSE_Results(1),RMSE_Results(2),RMSE_Results(3),RMSE_Results(4),RMSE_Results(5),RMSE_Results(6),RMSE_Results(7),...
            RMSE_Results(8),RMSE_Results(9),RMSE_Results(10),mean_RMSE_temp(1),mean_RMSE_temp(2),mean_RMSE_temp(3),mean_RMSE_temp(4),mean_RMSE_temp(5),mean_RMSE_temp(6),mean(RMSE_SOC_weighted),mean(RMSE_Sens_Offset_relevant)}';
        weightedScore(:,1) = Tests_weighted;                        % Load test names to table
        weightedScore(:,2) = Weights;                               % Load weights to table
        weightedScore(:,3) = Scores;                                % Load scores to table
        for j = 1:length(Tests_weighted)
            weightedScore(j,4) = {cell2mat(Weights(j)).*cell2mat(Scores(j))};% Load weighted scores to table            
        end
        summary = weightedScore;
        Final_Score = sum(weightedScore{:,4});

        %% Delete any failed figures from folder
        clear Cycle_Name; clear Amb_Name; clear File
        if fail == 1                                            
            delete('*.fig')
            clearvars -except rootfolder detect Inputs Data filenum fileFolder outputFolder Tests Tests_weighted Setups Temperatures Normal_cycles Custom_cycles
            return
        end
        %% Save figures and Error summary
        save('Error_Summary_Table','OutputData')                        % Create a copy to also .zip with figures
        DataFileName = fullfile(strcat(fileFolder,' Saved'), sprintf('Model_%d Error Summary.zip',filenum)); % Create name for new zip where figures will be stored
        zip(DataFileName,{'Error_Summary_Table.mat','*.fig', '*.svg'}); % Load the copy and all figures to the zip
        fileInfo = dir(DataFileName);
        MB_Size = fileInfo.bytes/1024/1024;
        if MB_Size > 25                                                 % If everything together is too large to email, split up the data and the figures
            DataFileName1 = fullfile(strcat(fileFolder,' Saved'), sprintf('Model_%d Error Summary 1 of 2.zip',filenum)); % Create name for new zip where figures will be stored
            DataFileName2 = fullfile(strcat(fileFolder,' Saved'), sprintf('Model_%d Error Summary 2 of 2.zip',filenum)); % Create name for new zip where figures will be stored
            zip(DataFileName1,{'Error_Summary_Table.mat'});             % Load the copy and all data to the zip
            zip(DataFileName2,{'*.fig', '*.svg'});                      % Load the copy and all figures to the zip
            delete(DataFileName)                                        % Delete big zip file
            fprintf('Figures were too large to send a single mail, the data was split up instead!\n')
        end
        delete('*.fig')                                                 % Delete the generated pictures
        delete('*.svg')                                                 % Delete the generated pictures
        delete('Error_Summary_Table.mat')                               % Delete the error summary copy
        cd(rootfolder)
        %% Create leaderboard entry
        dt = datetime('now');                                           % Find current date and time
        dt.Format = 'yyyy-MM-dd HH:mm:ss';                              % Reshape the format
        stats{1,1} = cellstr(dt);                                       % Save the necessary information                                                    
        stats{1,2} = Author_Name;
        stats{1,3} = Author_Affiliation;
        stats{1,4} = Model_Name;
        stats{1,5} = round(Final_Score,3);
        for j = 1:18
            stats{1,j+5} = round(cell2mat(Scores(j)),3);
        end
        stats{1,24} = round(RMSE_Results(1),3);
        stats{1,25} = round(mean(nonzeros(MAE)),3);
        stats{1,26} = round(mean(nonzeros(MAXE)),3);
        stats{1,27} = Complexity;
        stats{1,28} = filenum;

        %% Emailing Author With for Results
        fprintf('Updating the leaderboard...\n')                        
        if isfile(strcat(fileFolder,'\Leaderboard.csv'))
            stats2 = readtable(strcat(fileFolder,'\Leaderboard.csv'), 'VariableNamingRule', 'preserve');  % Read leaderboard file
            stats2 = removevars(stats2,{'Position'});                       % Remove old ranking numbers
            existingScores = single(stats2{:, 5:26});                       % Columns 6 to 26 correspond to the test scores
            Duplicate = any(ismember(existingScores, [stats{(5:26)}] , 'rows')); % Detect any identical scoring leaderboard entries
            if Duplicate
                title2_temp = {Model_Name{1},'is already on the leaderboard!'};
                Title2 = join(title2_temp);
                sendmail(Author_Email{1},Title2{1}, ...
                    'Dear Author, your submission has identical results to an existing entry. For this reason your submission is not added to the leaderboard. If you think this is a mistake, please contact us. Thank you!')
                temp = dir().folder;
                tempFolder = strcat(temp,"\Models Saved");
                movefile(fullfile(tempFolder,strcat('Model__',mat2str(filenum),'.zip')), ...
                fullfile(tempFolder,strcat('Model__',mat2str(filenum),'_DUPLICATE.zip')));
                clear temp; clear tempFolder;
                fprintf(2,'Submission has identical test results to an existing entry. Ignoring duplicate submission!\n');
                return
            else
                stats2 = vertcat(stats,stats2);                             % Add new stats to leaderboard as a new row
                stats2 = sortrows(stats2,[5,1]);                            % Sorts leaderboard by weighted error
                stats2.Position = (1:height(stats2))';                      % Add new ranking numbers
                check = fopen(strcat(fileFolder,'\Leaderboard.csv'), 'a');
                while check == -1
                    fprintf(2,'The file is open in Excel. Please close it! Trying again in 60 seconds.\n');
                    pause(60);
                    check = fopen(strcat(fileFolder,'\Leaderboard.csv'), 'a');
                end
                fclose(check);
                writetable(stats2,strcat(fileFolder,'\Leaderboard.csv'))    % Write to leaderboard.csv file
                for F = 1:height(stats2)                                    % Find the same submission time to find the rank
                    if dateshift(dt, 'start', 'second')==dateshift(stats2.Submission_Time(F), 'start', 'second') % If same date has been found
                        f = F;                                              % Set rank to row found
                    end
                end
            end
        else                                                            % If leaderboard does not yet exist
            stats = cell2table(stats);                                  % Create a table form stats
            stats.Position = 1;                                         % Give only entry ranking 1
            stats.Properties.VariableNames = {'Submission_Time','Author','Affiliation','Model_Name','Weighted_Error','Test_1: All_Cells','Test_2: Blind_Cells','Test_3: Non-Blinded_Cells','Test_4: Charging','Test_5: 80kg_Payload','Test_5-6: 448kg_Payload_with_HVAC', ...
                                        'Test_5-6: 448kg_Payload_no_HVAC','Test_5: 1000kg_Payload','Test_7: Standard_Cycles','Test_8: Custom_Cycles','Test_9: n20C','Test_9: n10C','Test_9: 0C','Test_9: 10C','Test_9: 25C','Test_9: 40C','Test_10: iSOC_Error',...
                                        'Test_11: Current_Sensor_Error','All_Drive_Cycles_Average_RMSE', 'All_Drive_Cycles_Average_MAE','All_Drive_Cycles_Average_MAXE','Complexity','Submission_ID' 'Position'};
            writetable(stats,strcat(fileFolder,'\Leaderboard.csv'));    % Write to leaderboard.csv file
            f = 1;
            stats2 = stats;
        end
        
        title2_temp = {Model_Name{1},'model results are ready!'};
        Title2 = join(title2_temp);
        lbpath = strcat(fileFolder,'\Leaderboard.csv');
        
        if exploit == 0                                     % If submission seems valid, send results and timedata
            if MB_Size < 25                                 % If full zip is small enough, send in one go
                emailMessage = sprintf(['Dear Author, Thank you for submitting your model. The Model currently ranks %d out of %d total submissions. ' ...
                    'The leaderboard,error results for all drive cycles and test result figures are attached to this email. Thank you!'], f, height(stats2));
                sendmail(Author_Email{1},Title2{1}, emailMessage, {DataFileName{1}, lbpath{1}});
            else                                            % If full zip is too big, send in parts
                emailMessage1 = sprintf(['Dear Author, Thank you for submitting your model. The Model currently ranks %d out of %d total submissions. ' ...
                    'The leaderboard and error results for all drive cycles are attached to this email. Due to the size of the data the test result figures are sent in another mail. Thank you!'], f, height(stats2));
                sendmail(Author_Email{1},Title2{1}, emailMessage1, {DataFileName1{1}, lbpath{1}});
                emailMessage2 = sprintf('Dear Author, Attached are the test result figures. Thank you!');
                sendmail(Author_Email{1},Title2{1}, emailMessage2, DataFileName2{1});
            end
        else                                                % If submission seems exploitatitve, send ranking but no data
            emailMessage = sprintf(['Dear Author, Thank you for submitting your model. The Model currently ranks %d out of %d total submissions. ' ...
                'Due to large deviations, no error data will be sent to you, try to improve your estimation results using the open data before submitting your model. ' ...
                'A possible reason for unexpected deviation could be that NaN values are returned for some drive cycles. ' ...
                ' If you think this is a mistake, please contact us. Thank you!'], f, height(stats2));
            sendmail(Author_Email{1},Title2{1}, emailMessage, lbpath{1});
        end
        else                                                % If there is no *.xlsx file in the correct location   
            title1_temp = {'Your model named "',Model_Name{1},'" has been submitted incorrectly!'}; % Add email title
            Title1 = join(title1_temp);
            sendmail(Author_Email{1},Title1{1}, ...         % Write and send email for submission failure
              'Dear Author, Your submission has been submitted incorrectly. No Model.m or Model.p file was found. Please resubmit your .zip with the correct names in the right format. Thank you!');
            fprintf("No Model.m or Model.p file found in folder, email sent!\n")
            delete(DataFileName)
            return
        end
        else
            fprintf(".xlsx file has the wrong format, no e-mail could be sent!\n")
            delete(DataFileName)
            return
        end
    else                                                            % If there is no *.xlsx file in the correct location   
        fprintf("No .xlsx file found in folder, no e-mail could be sent!\n")
        delete(DataFileName)
        return
    end                                                             % End for each model uploaded
    if fail == 0
        fprintf('The Model has been processed and ranks as #%d on the leaderboard.\n',f)
        close all; clearvars -except rootfolder detect Inputs Data filenum fileFolder outputFolder Tests Tests_weighted Setups Temperatures Normal_cycles Custom_cycles
    end
end